compiler/backend/src/org/jetbrains/kotlin/codegen/PropertyCodegen.java

```diff
@@ -208,8 +215,16 @@ private boolean isAccessorNeeded(
         // Delegated or extension properties can only be referenced via accessors
         if (declaration.hasDelegate() || declaration.getReceiverTypeReference() != null) return true;
 
-        // Companion object properties always should have accessors, because their backing fields are moved/copied to the outer class
-        if (isCompanionObject(descriptor.getContainingDeclaration())) return true;
+        // Companion object properties should have accessors for non-private properties because these properties can be referenced
+        // via getter/setter. But these accessors getter/setter are not required for private properties that have a default getter
+        // and setter, in this case, the property can always be accessed through the accessor 'access<property name>$cp' and avoid some
+        // useless indirection by using others accessors.
+        if (isCompanionObject(descriptor.getContainingDeclaration())) {
+            if (Visibilities.isPrivate(descriptor.getVisibility()) && isDefaultGetterAndSetter) {
```

Comment:
I'm a bit confused by this. Why is the fact that _both_ property accessors are default so important that it defines whether or not should we generate each of those accessors? The common sense is that the getter should be generated only if the getter is non-default, and the setter should be generated only if the setter is non-default. If only one of them is non-default, then that should be generated, and the other should not.

For example, in this case:
```
class A {
    companion object {
        private var x: String = ""
            get() = field
    }
}
```
I'd expect `A.Companion.getX` but _no setter_ to be generated.

I also couldn't find any tests on this behavior, since after dropping these lines (and relying on the accessor visibility check below in this method) all codegen tests still pass.

Reply:
Totally agree with you, it was my first attempt but solving this problem seems intrusive into the compiler, thus I go back to concentrate on the fix for KT-14258. 
My plan is to file another bug because as you mentioned, I think that the compiler can do again a better job by separating what is did for `getter` and `setter`, but it seems easier to work on it in a separate way to avoid to block the resolution of all cases without custom getter and setter.
One of the problem comes from the fact that the access to a property will not be always do in the same way, it depends on the context, nevertheless this behavior is not compliant with the usage of `accessors` map into `CodegenContext` indeed the key does not take into account if getter or setter accessors are required.